### ISSUE TITLE
Client Perf issues with Large Payload

### ISSUE THREAD

**Issue by rrichardson** [hyperium/hyper#1706]
tl;dr;  I suggest that we create a DEFAULT_BUFFER_SIZE that is much larger than the INIT_BUFFER_SIZE or better, we make buffer sizes configurable.  

I am building an object-storage benchmarker and I noticed that for large payloads (> 100MB) the rusoto/hyper client used 100% CPU per parallel core.  I traced it back to the fact that the hyper client io operates on buffers of 8k. 

I created a benchmark to validate this and found that if I am doing nothing with the Body stream, like the current end-to-end bench, but against 100MB payloads: 
```
8k buffers results in 1900MB/s  with a parallelism of 1
1MB buffers results in 4400MB/s with a paralleism of 1
```

If I actually do something with the buffer, for instance rusoto creates a new vec for each Chunk of the Body (that's another issue) the perf is as follows : 
```
8k buffers :   1600MB/s
1MB buffers: 3100MB/s
10MB buffers: 3300MB/s 
```
This is on Linux with a Intel(R) Core(TM) i7-6850K CPU @ 3.60GHz 

My benchmark code can be found here :  https://github.com/rrichardson/bench-body/blob/master/benches/bench.rs 

I've been going through this code for a while, and I'm not sure if I can pin down exactly where it picks how much it reads into a buffer,  since Buffered::read_buf is a BytesMut, it should be able to expand, but it never does, it stays at the initial size of 8192 bytes.  My guess is here : 
https://github.com/hyperium/hyper/blob/master/src/proto/h1/io.rs#L163   we  reserve INIT_BUFFER_SIZE and so it only reads in INIT_BUFFER_SIZE bytes?  Seems odd.

**Comment by seanmonstar** [comment]
Is it actually the buffer size passed to the socket? I've understood that TCP packets usually won't be more than 8kb unless on a local network...

**Comment by seanmonstar** [comment]
v0.12.16 has been released that includes a `http1_read_buf_exact_size` client option, allowing to set the read buffer to much higher.
